File: vision/pipeline.py
"""Vision orchestration: detection → tracking → events → (clips are handled by the API).

The core of the system. Returns a structure the API persists to the DB and uses to cut
clips. Designed as a pure domain function — no DB/HTTP dependencies.
"""
from collections.abc import Callable
from dataclasses import dataclass, field

from vision.clips import probe_video
from vision.detector import Detector
from vision.events import EventConfig, detect_events
from vision.types import DetectedEvent


@dataclass
class PipelineResult:
    duration_seconds: float | None
    fps: float | None
    events: list[DetectedEvent] = field(default_factory=list)
    frames_processed: int = 0


def analyze_video(
    video_path: str,
    *,
    yolo_model_path: str | None = None,
    frame_stride: int = 5,
    event_config: EventConfig | None = None,
    on_progress: Callable[[float], None] | None = None,
) -> PipelineResult:
    """Full recording analysis → detected events.

    on_progress: callback(0..1) for progress reporting (UI/queue).
    """
    duration, fps = probe_video(video_path)

    detector = Detector(model_path=yolo_model_path, frame_stride=frame_stride)

    frames = []
    for fr in detector.run(video_path):
        frames.append(fr)
        if on_progress and duration and fr.timestamp_seconds > 0:
            on_progress(min(0.95, fr.timestamp_seconds / duration))

    events: list[DetectedEvent] = detect_events(frames, event_config)

    if on_progress:
        on_progress(1.0)

    return PipelineResult(
        duration_seconds=duration,
        fps=fps,
        events=events,
        frames_processed=len(frames),
    )

File: worker-dotnet/PitchWise.Vision/Pipeline.cs
namespace PitchWise.Vision;

/// <summary>Result of a full recording analysis. Port of vision/pipeline.py PipelineResult.</summary>
public sealed class PipelineResult
{
    public double? DurationSeconds { get; init; }
    public double? Fps { get; init; }
    public IReadOnlyList<DetectedEvent> Events { get; init; } = Array.Empty<DetectedEvent>();
    public int FramesProcessed { get; init; }
}

/// <summary>
/// Vision orchestration: detection → tracking → events. Port of vision/pipeline.py
/// analyze_video. Pure domain function — no DB/HTTP dependencies.
/// </summary>
public static class Pipeline
{
    /// <param name="videoPath">Path to the recording.</param>
    /// <param name="modelPath">Exported YOLO11 .onnx.</param>
    /// <param name="classNames">model class-id → raw name.</param>
    /// <param name="frameStride">Process every Nth frame.</param>
    /// <param name="imgsz">Model input size (must match the export).</param>
    /// <param name="eventConfig">Event heuristics tuning (defaults if null).</param>
    /// <param name="onProgress">Progress callback (0..1) for UI/queue.</param>
    /// <param name="startSeconds">Analyse only from this timestamp (segment start). 0 = beginning.</param>
    /// <param name="endSeconds">Analyse only up to this timestamp (segment end). null = end of file.</param>
    /// <remarks>Events keep ABSOLUTE timestamps relative to the whole video, so a segmented
    /// run and a whole-file run place events at the same match time.</remarks>
    public static PipelineResult AnalyzeVideo(
        string videoPath,
        string modelPath,
        IReadOnlyDictionary<int, string> classNames,
        int frameStride = 5,
        int imgsz = 640,
        EventConfig? eventConfig = null,
        Action<double>? onProgress = null,
        double startSeconds = 0.0,
        double? endSeconds = null)
    {
        (double? duration, double? fps) = FfmpegTools.ProbeVideo(videoPath);

        int frameRate = (int)Math.Round(fps is > 0 ? fps.Value : 25.0);
        using var detector = new Detector(
            modelPath, classNames, frameRate: frameRate,
            frameStride: frameStride, imgsz: imgsz);

        // Progress is reported 0..1 over the requested [start, end] window (or whole file).
        double segStart = Math.Max(0.0, startSeconds);
        double segEnd = endSeconds ?? duration ?? 0.0;
        double segSpan = segEnd > segStart ? segEnd - segStart : 0.0;

        var frames = new List<FrameResult>();
        foreach (FrameResult fr in detector.Run(videoPath, startSeconds, endSeconds))
        {
            frames.Add(fr);
            if (onProgress is not null && segSpan > 0 && fr.TimestampSeconds > segStart)
                onProgress(Math.Min(0.95, (fr.TimestampSeconds - segStart) / segSpan));
        }

        List<DetectedEvent> events = Events.Detect(frames, eventConfig);

        onProgress?.Invoke(1.0);

        return new PipelineResult
        {
            DurationSeconds = duration,
            Fps = fps,
            Events = events,
            FramesProcessed = frames.Count,
        };
    }
}
